### ISSUE TITLE
Incorrect path matching when using nested paths

### ISSUE THREAD

**Issue by jwoudenberg** [router5/router5#53]
I'm running into some unexpected problems when defining nested paths. I was hoping you could set me on the right path (no pun intended :smile:). I have created a small example to demonstrate.

#### Problem
```js
var Router5 = require('router5').default;
var router = new Router5();

router
  .addNode('personList', '/persons/')
  .addNode('personDetail', '/persons/:personId');

var match = router.matchPath('/persons/jwoudenberg');
console.log(match); //=> null
```

In the example above, the path `/persons/jwoudenberg` does not result in a match. I would expect it to match the `personDetail` route.

#### Analsysis
After looking through the code I see what happens. The path is a partial match with the `personList` route. Router5 looks to complete the match by going through the child router of `personList`, but none exist, so it gives up. Removing the `personList` route makes `/persons/jwoudenberg` resolve correctly.

#### Solutions
I can imagine I could get it to work by defining `personDetail` as a child of `personList`. I was kind of hoping though that I could get around using this extra functionality, which I otherwise don't need. Certainly for new users, I think this can be confusing (as it was for me :smile:).

Some possible solutions I see:
1. Sort routes by the amount of segments they contain in descending order before matching. That way the most specific routes should be matched first.
2. Don't bail on the first failing partial match. If there's child routes left that haven't been tried yet, continue with those.

I'm looking forward to hearing your thoughts!

**Comment by troch** [comment]
I'll have a look, thanks for your detailed issue.

Normally route-node (which matches nested routes) sorts siblings so what you describe doesn't happen. Whether or not you define personList or personDetail first, route-node should attempt to match `personDetail` first.

Below is the logic:
https://github.com/troch/route-node/blob/master/modules/RouteNode.js#L96-L118

I'll run some more tests.

**Comment by jwoudenberg** [comment]
Thanks for responding quickly!

I stepped through the logic to figure out what was going on. I did (in my more complicated base case) notice that the routes were sorted in some way, but apparently not in the right order.

If it helps, you should be able to run the example in my post above as is and get the result I described. That did work for me (using the latest version of router5).

**Comment by troch** [comment]
Yeah looking at the logic, something didn't make sense. I'll need to add to my test cases and make sure everything still passes.

**Comment by jwoudenberg** [comment]
I found another issue, that is a bit hard to create a small example for basically, it deals with two routes like this:

```
postNew:    /blogs/:blogId/posts/new
postDetail: /blogs/:blogId/posts/:postId
```

Under certain circumstances, a path like `/blogs/myBlog/posts/new` will be matched to the `postDetail` route. It seems to occur seemingly randomly on the other routes defined.

I took a look in the code and noticed that these two routes score equal on all sorting criteria, which would mean their ordering depends on the complete list and the implementation of the browsers sorting algorithm. I would argue though, that the `postNew` route is clearly more specific, and so should always come first.

**Comment by tokenvolt** [comment]
I'm having the same issue. Currently I bypass this by defining different URL:
```js
.addNode('users', '/users-tab')
.addNode('user', '/users/:id')
```

**Comment by troch** [comment]
OK, I'll look into this this week.

**Comment by mephju** [comment]
I'd really like to see this fixed as well since it affects rather important routing functionality.

**Comment by troch** [comment]
Fixed in version 3.0.2.

There is one edge case left which requires a bit more work (along the lines of what @jwoudenberg mentioned with _Don't bail on the first failing partial match. If there's child routes left that haven't been tried yet, continue with those._). Considering @jwoudenberg example:

```
postNew:    /blogs/:blogId/posts/new
postDetail: /blogs/:blogId/posts/:postId
```

Everything works as long as `postId` doesn't start with `'new'`. Otherwise @jwoudenberg other case (with persons) and @tokenvolt example should be working fine.

For those interested: https://github.com/troch/route-node/blob/master/modules/RouteNode.js#L58-L83 for the improved sorting logic.

If any issues let me know. Thanks for your patience.

**Comment by jwoudenberg** [comment]
Works like a charm. Thanks for this!
